### PR TITLE
CPB-17: Vacancy model

### PR DESCRIPTION
**Pull request by LV-coding**

### CODE REVIEW

File: backend/apps/vacancies/admin.py
Comment (pazuzu-ua) [inline backend/apps/vacancies/admin.py]: You can put it in a decorator, looks cooler =)
```python
@admin.register(Vacancy)
class VacancyAdmin(admin.ModelAdmin):
```

File: backend/apps/vacancies/models.py
Comment (pazuzu-ua) [inline backend/apps/vacancies/models.py]: Is there any need to put it in float? It is usually counted in years or year periods. Maybe create another choices for it? Like: 0-1, 1-3... 5+ years.

File: backend/apps/vacancies/models.py
Comment (pazuzu-ua) [inline backend/apps/vacancies/models.py:12]: How about adding 'INTERNERSHIP' and 'VOLUNTEER'?

File: backend/apps/vacancies/models.py
Comment (pazuzu-ua) [inline backend/apps/vacancies/models.py:110]: Why is it a text field? Maybe put it in choices as well? If I correctly understand the meaning of the field, though...

**Review by pazuzu-ua — APPROVED** [review]

File: backend/apps/vacancies/models.py
Comment (devoorhees) [reply backend/apps/vacancies/models.py:12]: I also put these choice classes into a seperate file in the next pull request, so you need to change imports after or wait for my pull request to be merged first and then rebase your branch so you don't forget.

File: backend/apps/vacancies/models.py
Comment (devoorhees) [reply backend/apps/vacancies/models.py:110]: I also think the choices would be better in here.
Something like: "urgent, high, medium, low".

File: backend/apps/vacancies/models.py
Comment (devoorhees) [inline backend/apps/vacancies/models.py:145]: I don't think it's a good idea to validate the fields inside the `clean()` method because it will work only when you create a model instance.
The django documentation says: "a model’s clean() method is not invoked when you call your model’s save() method"
(https://docs.djangoproject.com/en/4.1/ref/models/instances/#django.db.models.Model.clean)
We should validate these fields inside the form or serializer.
(https://stackoverflow.com/a/36413343/15390079)

File: backend/apps/vacancies/models.py
Comment (devoorhees) [inline backend/apps/vacancies/models.py]: Did you mean `salary_min`?
Also make sure you mentioned all the class attributes in docstring.

File: backend/apps/vacancies/models.py
Comment (LV-coding) [reply backend/apps/vacancies/models.py:110]: @Voorhees2019 @pazuzu-ua 
As far as I understand, the priority is what is most important in the candidates, what will be an advantage among other candidates, etc.

File: backend/apps/vacancies/models.py
Comment (LV-coding) [reply backend/apps/vacancies/models.py:12]: > I also put these choice classes into a seperate file in the next pull request, so you need to change imports after or wait for my pull request to be merged first and then rebase your branch so you don't forget.

I thought, maybe later I will fix vacancies and other models with a hotfix

Comment (LV-coding) [reply backend/apps/vacancies/models.py:12]: > How about adding 'INTERNERSHIP' and 'VOLUNTEER'?

This is a slightly different grouping. EmploymentChoices is how often it will be necessary to work on this vacancy. INTERNERSHIP can be like fulltime or parttime. This is another category, closer to a grade.

VOLUNTEER also can be like fulltime or parttime.

File: backend/apps/accounts/migrations/0001_initial.py
Comment (devoorhees) [inline backend/apps/accounts/migrations/0001_initial.py]: Why did you delete migrations from the app you're not even working with??
Btw, you're not allowed to just delete migrations the other teammates already have.

File: backend/apps/vacancies/models.py
Comment (LV-coding) [reply backend/apps/vacancies/models.py]: > Is there any need to put it in float? It is usually counted in years or year periods. Maybe create another choices for it? Like: 0-1, 1-3... 5+ years.

This is good for analytics, but in the case of job vacancies, it can make the reach too large. Artificially increase the number of reviews for the vacancy by extra people.
For example, the difference between 1 year of experience and 3 years is huge. Both in terms of skills and salary.

File: backend/apps/vacancies/models.py
Comment (devoorhees) [reply backend/apps/vacancies/models.py:110]: @LV-coding 
Maybe I didn't get a thing...
I thought that was a priority of hiring after a review or something similar.

File: backend/apps/accounts/migrations/0001_initial.py
Comment (LV-coding) [reply backend/apps/accounts/migrations/0001_initial.py]: > Why did you delete migrations from the app you're not even working with?? Btw, you're not allowed to just delete migrations the other teammates already have.

Because precommit don't miss it. 
.gitignor should not skip migrations.

File: backend/apps/vacancies/models.py
Comment (devoorhees) [reply backend/apps/vacancies/models.py]: @LV-coding 
You can make choices "1, 2, 3... years" because, I guess, noone's specifying their experience like 1.2 or 2.7 years.

File: backend/apps/accounts/migrations/0001_initial.py
Comment (devoorhees) [reply backend/apps/accounts/migrations/0001_initial.py]: First of all, please revert changes and return those migrations back.
Then find out where the problem came from (which tool is exactly removing the migration files), because we had excluded the migrations inside the `pre-commit-config.yaml` file.
And .gitignore is not staging migrations, what's actually right.

File: backend/apps/vacancies/models.py
Comment (LV-coding) [reply backend/apps/vacancies/models.py:145]: But we can simply call the clean method already in the serializer

Comment (devoorhees) [reply backend/apps/vacancies/models.py:145]: Such way will be implicit for the whole team because everyone must remember to call `model.clean()` method all the time they need to `save` the instance inside all of the serializer's methods or anywhere else...
Serializers have their own `validate()` method which is exactly designed for this.
Keep it simple! Imagine if you leave the project, a new developer comes on a board and he/she has no idea about that specific rule (always call `clean()` method before saving a vacancy instance). That is very inconvenient and leads to a lot of problems.

File: backend/apps/vacancies/models.py
Comment (pazuzu-ua) [reply backend/apps/vacancies/models.py:110]: Maybe link it to some field in Candidate model then? Create some model 'Qualities' or something.

**Pull request merged**
